### PR TITLE
Improve performance issue by reducing processing rows

### PR DESCRIPTION
**Pull request by yuki-takei**
### Initial checklist

*   [x] I read the support docs 
*   [x] I read the contributing guide 
*   [x] I agree to follow the code of conduct 
*   [x] I searched issues and couldn’t find anything (or linked relevant results below) 
*   [ ] If applicable, I’ve added docs and tests

### Summary

This is a proposal-level PR for resolving performance issue #6.
And there are some duplicate lists.

- https://github.com/micromark/micromark-extension-gfm-table/issues/9
- https://github.com/micromark/micromark-extension-gfm/issues/8
- https://github.com/remarkjs/remark-gfm/issues/45
- https://github.com/orgs/remarkjs/discussions/978

### Investigation

We can display the method calling counts by adding the below code on the beginning of `atRowEndHead`:

```javascript
const { containerState } = self;
let atRowEndHeadCount = containerState.atRowEndHeadCount ?? 0;
atRowEndHeadCount++;
containerState.atRowEndHeadCount = atRowEndHeadCount;
console.log({ atRowEndHeadCount });
```

|markdown| count (original code) |
|---|---:|
|\| heading \|<br>\| - \||2|
|\| heading  \|<br>\| - \|<br>\| body1 \||4|
|\| heading  \|<br>\| - \|<br>\| body1 \|<br>\| body2 \||7|
|\| heading  \|<br>\| - \|<br>\| body1 \|<br>\| body2 \|<br>\| body3 \||11|
| ... | | |
|\| heading  \|<br>\| - \|<br>\| body1 \|<br>\| body2 \|<br>\| ... \|<br>\| bodyN \|| (2+N) + (1 + 2 + ... + N) |

Though `cellDividerHead` and `inCellContentHead` should be no longer needed when processing body rows, these are called repeatedly for the number of body rows. As a result, counts are increased to `allRows.length + 1 + 2 + 3 + ... + bodyRows.length`.

Furthermore, if the table includes delimiter-like rows in body, the count will explode.

|markdown| count (original code) |
|---|---:|
|\| heading \|<br>\| ------- \|<br>\| body1   \|<br>\| body2   \|<br>\| body3   \||11|
|\| heading \|<br>\| ------- \|<br>\| body1   \|<br>\| body2   \|<br>\| body3   \|<br>\| ------- \|<br>\| body1   \|<br>\| body2   \|<br>\| body3   \||62|
|\| heading \|<br>\| ------- \|<br>\| body1   \|<br>\| body2   \|<br>\| body3   \|<br>\| ------- \|<br>\| body1   \|<br>\| body2   \|<br>\| body3   \|<br>\| ------- \|<br>\| body1   \|<br>\| body2   \|<br>\| body3   \||273|
|\| heading \|<br>\| ------- \|<br>\| body1   \|<br>\| body2   \|<br>\| body3   \|<br>\| ------- \|<br>\| body1   \|<br>\| body2   \|<br>\| body3   \|<br>\| ------- \|<br>\| body1   \|<br>\| body2   \|<br>\| body3   \|<br>\| ------- \|<br>\| body1   \|<br>\| body2   \|<br>\| body3   \||1124|

I have not investigated what mechanism causes this.

### Description of changes

I modified `start` function to skip calling `cellDividerHead(code)` and `inCellContentHead(code)` by counting row counts processed. 
This fixes `atRowEndHeadCount` to 2 times. This is the minimum processing for the heading row and the delimiter row.

### Note

This PR code is proposal-level and not for production.

- I have defined `rowCount` and `hasDelimiterRow` in `self.containerState` so that the information is persisted throughout the flow of processing the table.
    - I didn't have no idea about what `tableHeaderCount`, `seenDelimiter` and `hasDash` were for. Should I use it? But these are initialized every `start`.
- I have been recognized the test will fail for `should *not* be interrupted by a heading (setext), but interrupt if the underline is also an empty list item bullet`.
    - I will add some line comments for this after now.

I'm happy if the PR will be some hints to solve the problem.

### CODE REVIEW

File: dev/lib/syntax.js [hunk 1]
```diff
@@ -176,11 +176,24 @@ function tokenizeTable(effects, ok, nok) {
...
     effects.enter('table')._align = align
     effects.enter('tableHead')
     effects.enter('tableRow')
 
+    // increment row count
+    const crrRowCount = prevRowCount + 1;
+    containerState.rowCount = crrRowCount;
+
+    // Max 2 rows processing before delimiter row
+    if (hasDelimiterRow || crrRowCount > 2) {
```
Comment (yuki-takei) [inline dev/lib/syntax.js:193]: All tests will pass if you change the condition to `hasDelimiterRow || crrRowCount > 3`.

Because the given markdown int the test:

```markdown
| a |
| - |
heading
-
```

needs 3-rows processing.

However, Is this check required for the implementation of common-mark or GFM? I think it'll be not bad to omit the test if it is a corner case.

**Comment by yuki-takei** [comment]
I investigated the new code the same way and how many `headRowBreak` was called.
It was called just 6 times in any input case.

The performance issue has been solved completely. Thanks.
